### PR TITLE
[SPARK-10359] Enumerate dependencies in a file and diff against it for new pull requests

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch adds a new build check which enumerates Spark's resolved runtime classpath and saves it to a file, then diffs against that file to detect whether pull requests have introduced dependency changes. The aim of this check is to make it simpler to reason about whether pull request which modify the build have introduced new dependencies or changed transitive dependencies in a way that affects the final classpath.

This supplants the checks added in SPARK-4123 / #5093, which are currently disabled due to bugs.

This patch is based on @pwendell's work in #8531.

Closes #8531.

### CODE REVIEW

File: dev/deps/spark-deps-hadoop-2.4 [hunk 1]
```diff
@@ -0,0 +1,185 @@
+JavaEWAH-0.3.2.jar
```
Comment (nchammas) [inline dev/deps/spark-deps-hadoop-2.4]: Do we have to maintain these lists? Can't we use Maven to generate them from master, generate from PR, and compare?

Or is that an expensive operation?

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: dev/deps/spark-deps-hadoop-2.4 (hunk 1 above)
Comment (JoshRosen) [inline dev/deps/spark-deps-hadoop-2.4]: IMO, there's very little overhead to maintaining these lists in the repo and I find it helpful to explicitly see the diff of the dependencies via `git`.

Comment (pwendell) [inline dev/deps/spark-deps-hadoop-2.4]: yes these are automatically generated, so it's not a huge maintenance cost. I think having them int he repo is good so people can have a definitive reference for what dependencies exist in which package of spark.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
LGTM

**Comment by JoshRosen** [comment]
Let me try introducing a dependency change just to confirm that it works end-to-end before I merge...

**Comment by yhuai** [comment]
test this please

**Comment by JoshRosen** [comment]
Let me try that dummy change again...

**Comment by JoshRosen** [comment]
Uh oh, I'd hold off on merging this pending investigation of issues with the dummy change.

**Comment by JoshRosen** [comment]
Spotted the problem: the original logic for checking `should_run_build_tests` was wrong: it needs to loop over all `test_modules` (using `any`) and the `root` module needs to have the build tests enabled.

**Comment by JoshRosen** [comment]
Woohoo, that worked! However, I'm not a fan of how vanilla `diff` formats its output, so I'm going to quickly try and see if I can use `git diff` for this.

**Comment by JoshRosen** [comment]
Going to merge this into master and will email the dev list with a summary describing the changes + impact on process.

**Pull request closed without merging**